tests/test_qp.py

```diff
@@ -22,317 +24,245 @@ class QPTest(PyverbsAPITestCase):
-                        qp = QP(ctx, qia)
-                    except PyverbsRDMAError as ex:
-                        if ex.error_code == errno.EOPNOTSUPP:
-                            raise unittest.SkipTest('Create QP with extended attrs is not supported')
-                        raise ex
+                    qp = self.create_qp(ctx, qia, False, True)
                     qa = QPAttr()
                     qa.qkey = 0x123
                     qp.to_init(qa)
-                    qp_attr, qp_iattr = qp.query(e.IBV_QP_QKEY)
+                    qp_attr, _ = qp.query(e.IBV_QP_QKEY)
                     assert qp_attr.qkey == qa.qkey, 'Extended QP, QKey is not as expected'
                     qp.to_rtr(qa)
                     qa.sq_psn = 0x45
                     qp.to_rts(qa)
-                    qp_attr, qp_iattr = qp.query(e.IBV_QP_SQ_PSN)
+                    qp_attr, _ = qp.query(e.IBV_QP_SQ_PSN)
                     assert qp_attr.sq_psn == qa.sq_psn, 'Extended QP, SQ PSN is not as expected'
                     qa.qp_state = e.IBV_QPS_RESET
                     qp.modify(qa, e.IBV_QP_STATE)
                     assert qp.qp_state == e.IBV_QPS_RESET, 'Extended QP, QP state is not as expected'
                     # Legacy QP
                     qia = get_qp_init_attr(cq, attr)
-                    qp = QP(pd, qia)
+                    qp = self.create_qp(pd, qia, False, False)
                     qa = QPAttr()
                     qa.qkey = 0x123
                     qp.to_init(qa)
-                    qp_attr, qp_iattr = qp.query(e.IBV_QP_QKEY)
+                    qp_attr, _ = qp.query(e.IBV_QP_QKEY)
                     assert qp_attr.qkey == qa.qkey, 'Legacy QP, QKey is not as expected'
                     qp.to_rtr(qa)
                     qa.sq_psn = 0x45
                     qp.to_rts(qa)
-                    qp_attr, qp_iattr = qp.query(e.IBV_QP_SQ_PSN)
+                    qp_attr, _ = qp.query(e.IBV_QP_SQ_PSN)
                     assert qp_attr.sq_psn == qa.sq_psn, 'Legacy QP, SQ PSN is not as expected'
                     qa.qp_state = e.IBV_QPS_RESET
                     qp.modify(qa, e.IBV_QP_STATE)
                     assert qp.qp_state == e.IBV_QPS_RESET, 'Legacy QP, QP state is not as expected'
 
 
-def get_qp_types(ctx, port_num):
-    """
-    Returns a list of the commonly used QP types. Raw Packet QP will not be
-    included if link layer is not Ethernet or it current user is not root.
-    :param ctx: The device's Context, to query the port's link layer
-    :param port_num: Port number to query
-    :return: An array of QP types that can be created on this port
-    """
-    qpts = [e.IBV_QPT_RC, e.IBV_QPT_UC, e.IBV_QPT_UD]
-    if u.is_eth(ctx, port_num) and u.is_root():
-        qpts.append(e.IBV_QPT_RAW_PACKET)
-    return qpts
-
-
 def verify_qp_attrs(orig_cap, state, init_attr, attr):
     assert state == attr.cur_qp_state
     assert orig_cap.max_send_wr <= init_attr.cap.max_send_wr
```

Comment:
btw you can move this to QPTest class, so you can use self.assertEqual instead.
i.e: self.assertEqual(state , attr.cur_qp_state).
This will print both values in case of failure (which assert doesn't do).

Reply:
Ack, moved to the QPTest class.